# Setup

In [1]:
# base
import matplotlib.pyplot as plt
plt.style.use('dark_background')
import seaborn as sns
import pandas as pd
import numpy as np
import datetime

In [2]:
import sys
sys.path.insert(1, '/home/mauricio/code/mcr')
from mcr.util import glimpse, plot_value_counts, plot_value_counts_timeseries, missing_report, plot_missing, plot_unique, plot_duplicates, size
from mcr.spark import roem

In [3]:
from pyspark import SparkContext
# SparkContext.getOrCreate(conf: Optional[pyspark.conf.SparkConf] = None) -> 'SparkContext'
SparkContext.setSystemProperty('spark.executor.memory', '2g')
SparkContext.setSystemProperty('spark.executor.cores', '16')
sc = SparkContext.getOrCreate()
# sc.setLogLevel('DEBUG')
# https://intellipaat.com/community/18452/spark-gives-a-stackoverflowerror-when-training-using-als
sc.setCheckpointDir('checkpoint/')
sc.getConf().getAll()

23/05/08 12:52:27 WARN Utils: Your hostname, rig resolves to a loopback address: 127.0.1.1; using 192.168.0.102 instead (on interface enp6s0)
23/05/08 12:52:27 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/05/08 12:52:27 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/05/08 12:52:28 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.


[('spark.executor.memory', '2g'),
 ('spark.executor.id', 'driver'),
 ('spark.driver.host', '192.168.0.102'),
 ('spark.app.name', 'pyspark-shell'),
 ('spark.app.id', 'local-1683561148303'),
 ('spark.driver.extraJavaOptions',
  '-Djava.net.preferIPv6Addresses=false -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -Djdk.refl

In [4]:
from pyspark.sql import functions as F
from pyspark.sql.types import *
from pyspark.sql import SparkSession
spark = SparkSession.builder.master('local[*]').appName('spark_application').getOrCreate()
print(spark.version)

3.4.0


# Introduction to the Million Songs Dataset

Thierry Bertin-Mahieux, Daniel P.W. Ellis, Brian Whitman, and Paul Lamere. The Million Song Dataset. In Proceedings of the 12th International Society for Music Information Retrieval Conference (SIMIR 2012), 2011.

You're going to be using one file of this dataset called The Echo Nest Taste profile dataset. It contains information on over 1 million users including the number of times they've played nearly 400,000 songs. This is more data than we can use for this course, so we will only be using a portion of.

We'll first examine the data, get summary statistics, and then build and evaluate our model. One thing to note here is that because the use of implicit ratings causes ALS to calculate a level of confidence that a user likes a song based on the number of times they've played it, the matrix will need to include zeros for the songs that each user has not yet listened to. In case your data doesn't already include the zeros, we'll walk through how to do this.

In [5]:
msd = spark.read.csv('msd_ratings_w_zeros_16JUL2018.csv',
                     sep=',',
                     header=True,
                     inferSchema=True)
msd.printSchema()
msd.show(3)
msd.count()

root
 |-- userId: integer (nullable = true)
 |-- songId: integer (nullable = true)
 |-- num_plays: integer (nullable = true)

+------+------+---------+
|userId|songId|num_plays|
+------+------+---------+
|   148|   148|        0|
|   243|   496|        0|
|    31|   471|        0|
+------+------+---------+
only showing top 3 rows



234009

In [6]:
# Why msd has duplicated rows by userId, songId that was not consolidated reducing rows from 234009 fot 78003 
# Why without consolidation the ROEM is improved?
# msd = msd.groupby(['userId', 'songId']).agg({'num_plays': 'sum'}).withColumnRenamed('sum(num_plays)', 'num_plays')

In [7]:
msd.show(3)
msd.count()

+------+------+---------+
|userId|songId|num_plays|
+------+------+---------+
|   148|   148|        0|
|   243|   496|        0|
|    31|   471|        0|
+------+------+---------+
only showing top 3 rows



234009

## Cross join intro

In [8]:
users = msd.select("userId").distinct().orderBy('userId')
users.show(3)
users.count()

+------+
|userId|
+------+
|     0|
|     1|
|     2|
+------+
only showing top 3 rows



321

In [9]:
songs = msd.select("songId").distinct().orderBy('songId')
songs.show(3)
songs.count()

+------+
|songId|
+------+
|     0|
|     1|
|     2|
+------+
only showing top 3 rows



729

## Cross join output

Returns the cartesian product with another DataFrame

In [10]:
cross_join = users.crossJoin(songs)
cross_join.show(3)
cross_join.count()

+------+------+
|userId|songId|
+------+------+
|   148|   148|
|   243|   148|
|    31|   148|
+------+------+
only showing top 3 rows



234009

## Joining back original ratings data

In [11]:
msd = cross_join\
    .join(msd, ["userId", "songId"], "left").orderBy(['userId', 'songId'])
msd.show(3)
msd.count()

+------+------+---------+
|userId|songId|num_plays|
+------+------+---------+
|     0|     0|        2|
|     0|     0|        2|
|     0|     0|        2|
+------+------+---------+
only showing top 3 rows



390015

## Filling in with zero

In [12]:
msd = msd.fillna(0, subset='num_plays').persist()
msd.show(3)

+------+------+---------+
|userId|songId|num_plays|
+------+------+---------+
|     0|     0|        2|
|     0|     0|        2|
|     0|     0|        2|
+------+------+---------+
only showing top 3 rows



## Add zeros function

In [13]:
def add_zeros(df, user_col, item_col):
    # Extracts distinct users
    users = df.select(user_col).distinct()
    # Extracts distinct songs
    songs = df.select(item_col).distinct()
    # Joins users and songs, fills blanks with 0
    cross_join = users.crossJoin(songs) \
    .join(df, [user_col, item_col], "left").fillna(0)
    return cross_join

## Exercises

### MSD summary statistics

In [14]:
# Look at the data
msd.show(3)
# Count the number of distinct userIds
user_count = msd.select("userId").distinct().count()
print("Number of users: ", user_count)
# Count the number of distinct songIds
song_count = msd.select("songId").distinct().count()
print("Number of songs: ", song_count)

+------+------+---------+
|userId|songId|num_plays|
+------+------+---------+
|     0|     0|        2|
|     0|     0|        2|
|     0|     0|        2|
+------+------+---------+
only showing top 3 rows

Number of users:  321


Number of songs:  729


In [15]:
# Min num implicit ratings for a song
print("Minimum implicit ratings for a song: ")
msd.filter(F.col("num_plays") > 0).groupBy("songId").count().select(F.min("count")).show()

# Avg num implicit ratings per songs
print("Average implicit ratings per song: ")
msd.filter(F.col("num_plays") > 0).groupBy("songId").count().select(F.avg("count")).show()

# Min num implicit ratings from a user
print("Minimum implicit ratings from a user: ")
msd.filter(F.col("num_plays") > 0).groupBy("userId").count().select(F.min("count")).show()

# Avg num implicit ratings for users
print("Average implicit ratings per user: ")
msd.filter(F.col("num_plays") > 0).groupBy("userId").count().select(F.avg('count')).show()

Minimum implicit ratings for a song: 
+----------+
|min(count)|
+----------+
|         3|
+----------+

Average implicit ratings per song: 
+------------------+
|        avg(count)|
+------------------+
|35.251063829787235|
+------------------+

Minimum implicit ratings from a user: 
+----------+
|min(count)|
+----------+
|        21|
+----------+

Average implicit ratings per user: 
+-----------------+
|       avg(count)|
+-----------------+
|77.42056074766356|
+-----------------+



> Users have at least 21 implicit ratings with an average of 77 and each song has at least 3 implicit ratings with an average of 35.

### Add zeros

In [16]:
Z = pd.DataFrame(
    {
        "num_purchases": {0: 1, 1: 23, 2: 9, 3: 2, 4: 5, 5: 21, 6: 8, 7: 96},
        "productId": {
            0: 777,
            1: 44,
            2: 227,
            3: 1981,
            4: 2390,
            5: 1662,
            6: 1492,
            7: 1811,
        },
        "userId": {0: 2112, 1: 7, 2: 1132, 3: 686, 4: 42, 5: 13, 6: 2112, 7: 22},
    }
)
Z = spark.createDataFrame(Z)
Z.show()
Z.count()

+-------------+---------+------+
|num_purchases|productId|userId|
+-------------+---------+------+
|            1|      777|  2112|
|           23|       44|     7|
|            9|      227|  1132|
|            2|     1981|   686|
|            5|     2390|    42|
|           21|     1662|    13|
|            8|     1492|  2112|
|           96|     1811|    22|
+-------------+---------+------+



8

In [17]:
# Extract distinct userIds and productIds
users = Z.select("userId").distinct()
products = Z.select("productId").distinct()

# Cross join users and products
cj = users.crossJoin(products)

# Join cj and Z
Z_expanded = cj.join(Z, ["userId", "productId"], "left").fillna(0)

# View Z_expanded
Z_expanded.show()
Z_expanded.count()

+------+---------+-------------+
|userId|productId|num_purchases|
+------+---------+-------------+
|    13|      777|            0|
|    42|      227|            0|
|     7|      227|            0|
|  1132|      777|            0|
|  2112|      777|            1|
|    22|      227|            0|
|    13|       44|            0|
|   686|      227|            0|
|    22|      777|            0|
|    42|      777|            0|
|   686|       44|            0|
|     7|       44|           23|
|  2112|       44|            0|
|   686|      777|            0|
|     7|      777|            0|
|  2112|      227|            0|
|    22|       44|            0|
|  1132|      227|            9|
|    42|       44|            0|
|  1132|       44|            0|
+------+---------+-------------+
only showing top 20 rows



56

# Evaluating implicit ratings models

## Evaluating implicit ratings models
Now that we have an implicit ratings dataset, let's discuss these types of models. The first thing you should know is that implicit ratings models have an additional hyperparameter called alpha. Alpha is an integer value that tells Spark how much each additional song play should add to the model's confidence that a user actually likes a song. Like the other hyperparameters, this will need to be tuned through cross validation. The challenge of these models is the evaluation. With explicit ratings, we used was the RMSE. It made sense in that situation because we could match predictions back to a true measure of user preference. In the case of implicit ratings.

## Why RMSE worked before

Predictions reflect actual ratings, RMSE makes sense here.

## Why RMSE doesn't work now

Different metrics, RMSE doesn't make sense here.

We don’t have a true measure of user preference. We only have the number of times a user listened to a song and a measure of how confident our model is that that they like that song. These aren't the same thing and calculating an RMSE between them doesn't make sense. However, using a test set, we can see if our model is giving high predictions to the songs that users have actually listened to. The logic being that if our model is returning a high prediction for a song that the respective user has actually listened to, then the predictions make sense, especially if they've listened to it more than once. We can measure this using the **Rank Ordering Error Metric (ROEM)** 

## (ROEM) Rank Ordering Error Metric

$$ROEM = \frac{\sum_{u,i} r_{u,i}^t rank_{u,i}}{\sum_{u,i} r_{u,i}^t}$$

>  Unfortunately Spark hasn't implemented an evaluator metric like this, so you'll need to build it manually. An ROEM function will be provided to you in subsequent exercises.

## ROEM bad predictions

In [18]:
bad_predictions = spark.createDataFrame(pd.DataFrame(
    columns=["userId", "songId", "plays", "prediction"],
    data=[
        [111, 22, 3, 0.0001],
        [111, 9, 0, 0.999],
        [111, 321, 0, 0.08],
        [222, 84, 0, 0.000003],
        [222, 821, 2, 0.88],
        [222, 91, 2, 0.73],
        [333, 2112, 0, 0.90],
        [333, 42, 2, 0.80],
        [333, 6, 0, 0.01],
    ],
)).sort(['userId', 'prediction'], ascending=[True, False])
bad_predictions.show()

+------+------+-----+----------+
|userId|songId|plays|prediction|
+------+------+-----+----------+
|   111|     9|    0|     0.999|
|   111|   321|    0|      0.08|
|   111|    22|    3|    1.0E-4|
|   222|   821|    2|      0.88|
|   222|    91|    2|      0.73|
|   222|    84|    0|    3.0E-6|
|   333|  2112|    0|       0.9|
|   333|    42|    2|       0.8|
|   333|     6|    0|      0.01|
+------+------+-----+----------+



### Percent Rank

In [19]:
bad_predictions.createOrReplaceTempView("predictions")
#Calculating rankings of songs predictions by user
spark.sql("SELECT userId, songId, plays, prediction, PERCENT_RANK() OVER (PARTITION BY userId ORDER BY prediction DESC) AS percRank FROM predictions")\
     .createOrReplaceTempView("rankings")
spark.sql('SELECT * FROM rankings').show()

+------+------+-----+----------+--------+
|userId|songId|plays|prediction|percRank|
+------+------+-----+----------+--------+
|   111|     9|    0|     0.999|     0.0|
|   111|   321|    0|      0.08|     0.5|
|   111|    22|    3|    1.0E-4|     1.0|
|   222|   821|    2|      0.88|     0.0|
|   222|    91|    2|      0.73|     0.5|
|   222|    84|    0|    3.0E-6|     1.0|
|   333|  2112|    0|       0.9|     0.0|
|   333|    42|    2|       0.8|     0.5|
|   333|     6|    0|      0.01|     1.0|
+------+------+-----+----------+--------+



### Number of plays X rank

In [20]:
spark.sql('SELECT *, plays*percRank as `np*rank` FROM rankings').createOrReplaceTempView("rankings")
spark.sql('SELECT * FROM rankings').show()

+------+------+-----+----------+--------+-------+
|userId|songId|plays|prediction|percRank|np*rank|
+------+------+-----+----------+--------+-------+
|   111|     9|    0|     0.999|     0.0|    0.0|
|   111|   321|    0|      0.08|     0.5|    0.0|
|   111|    22|    3|    1.0E-4|     1.0|    3.0|
|   222|   821|    2|      0.88|     0.0|    0.0|
|   222|    91|    2|      0.73|     0.5|    1.0|
|   222|    84|    0|    3.0E-6|     1.0|    0.0|
|   333|  2112|    0|       0.9|     0.0|    0.0|
|   333|    42|    2|       0.8|     0.5|    1.0|
|   333|     6|    0|      0.01|     1.0|    0.0|
+------+------+-----+----------+--------+-------+



### ROEM calculation

In [21]:
numerator = spark.sql('SELECT SUM(plays * percRank) FROM rankings').collect()[0][0]
denominator = bad_predictions.groupBy().sum('plays').collect()[0][0]
numerator / denominator

0.5555555555555556

In [22]:
roem(spark, bad_predictions, userCol='userId', ratingCol='plays')

0.5555555555555556

## ROEM good predictions

In [23]:
good_predictions = spark.createDataFrame(pd.DataFrame(
    columns=['userId' ,'songId','plays','prediction'],
    data=[[111, 22, 3, 1.1],
          [111, 77, 0, 0.01],
          [111, 99, 0, 0.008],
          [222, 22, 0, 0.0003],
          [222, 77, 2, 1.5],
          [222, 99, 2, 1.4],
          [333, 22, 0, 0.90],
          [333, 77, 2, 1.6],
          [333, 99, 0, 0.01]]))
good_predictions.show()

+------+------+-----+----------+
|userId|songId|plays|prediction|
+------+------+-----+----------+
|   111|    22|    3|       1.1|
|   111|    77|    0|      0.01|
|   111|    99|    0|     0.008|
|   222|    22|    0|    3.0E-4|
|   222|    77|    2|       1.5|
|   222|    99|    2|       1.4|
|   333|    22|    0|       0.9|
|   333|    77|    2|       1.6|
|   333|    99|    0|      0.01|
+------+------+-----+----------+



### Percent Rank

In [24]:
good_predictions.createOrReplaceTempView("predictions")
#Calculating rankings of songs predictions by user
spark.sql("SELECT userId, songId, plays, prediction, PERCENT_RANK() OVER (PARTITION BY userId ORDER BY prediction DESC) AS percRank FROM predictions")\
     .createOrReplaceTempView("rankings")
spark.sql('SELECT * FROM rankings').show()

+------+------+-----+----------+--------+
|userId|songId|plays|prediction|percRank|
+------+------+-----+----------+--------+
|   111|    22|    3|       1.1|     0.0|
|   111|    77|    0|      0.01|     0.5|
|   111|    99|    0|     0.008|     1.0|
|   222|    77|    2|       1.5|     0.0|
|   222|    99|    2|       1.4|     0.5|
|   222|    22|    0|    3.0E-4|     1.0|
|   333|    77|    2|       1.6|     0.0|
|   333|    22|    0|       0.9|     0.5|
|   333|    99|    0|      0.01|     1.0|
+------+------+-----+----------+--------+



### Number of plays X rank

In [25]:
spark.sql('SELECT *, plays*percRank as `np*rank` FROM rankings').createOrReplaceTempView("rankings")
spark.sql('SELECT * FROM rankings').show()

+------+------+-----+----------+--------+-------+
|userId|songId|plays|prediction|percRank|np*rank|
+------+------+-----+----------+--------+-------+
|   111|    22|    3|       1.1|     0.0|    0.0|
|   111|    77|    0|      0.01|     0.5|    0.0|
|   111|    99|    0|     0.008|     1.0|    0.0|
|   222|    77|    2|       1.5|     0.0|    0.0|
|   222|    99|    2|       1.4|     0.5|    1.0|
|   222|    22|    0|    3.0E-4|     1.0|    0.0|
|   333|    77|    2|       1.6|     0.0|    0.0|
|   333|    22|    0|       0.9|     0.5|    0.0|
|   333|    99|    0|      0.01|     1.0|    0.0|
+------+------+-----+----------+--------+-------+



### ROEM calculation

In [26]:
numerator = spark.sql('SELECT SUM(plays * percRank) FROM rankings').collect()[0][0]
denominator = good_predictions.groupBy().sum('plays').collect()[0][0]
numerator / denominator

0.1111111111111111

In [27]:
roem(spark, good_predictions, userCol='userId', ratingCol='plays')

0.1111111111111111

## Building several ROEM models

In [28]:
from pyspark.ml.recommendation import ALS

(train, test) = msd.randomSplit([.8, .2])

In [29]:
# Empty list to be filled with models
model_list = []
# Complete each of the hyperparameter value lists
ranks = [10, 20, 30, 40]
maxIters = [10, 20, 30, 40]
regParams = [.05, .1, .15]
alphas = [20, 40, 60, 80]
# For loop will automatically create and store ALS models
for r in ranks:
    for mi in maxIters:
        for rp in regParams:
            for a in alphas:
                model_list.append(ALS(userCol= "userId", itemCol= "songId",
                                      ratingCol= "num_plays", rank = r, maxIter = mi, regParam = rp,
                                      alpha = a, coldStartStrategy="drop",nonnegative = True,
                                      implicitPrefs = True))
print(len(model_list))

192


## Error output

In [30]:
ROEMS = []
for i, model in enumerate(model_list):
    # Fits each model to the training data
    print(f'Training model #{i} ', end='')
    trained_model = model.fit(train)
    # Generates test predictions
    predictions = trained_model.transform(test)
    # Evaluates each model's performance
    ROEMS.append(roem(spark, predictions, 'userId', 'num_plays'))
    print(f'ROEM:', ROEMS[-1])

Training model #0 ROEM: 0.08874493046678647
Training model #1 ROEM: 0.09663821944522209
Training model #2 ROEM: 0.10328372095011613
Training model #3 ROEM: 0.10724205888235687
Training model #4 ROEM: 0.08848094911677111
Training model #5 ROEM: 0.0957287518738229
Training model #6 ROEM: 0.1017360736868199
Training model #7 ROEM: 0.10580659792092394
Training model #8 ROEM: 0.08738962956366758
Training model #9 ROEM: 0.09497871384949229
Training model #10 ROEM: 0.1002834323182076
Training model #11 ROEM: 0.10478079704851734
Training model #12 ROEM: 0.0883444365544352
Training model #13 ROEM: 0.09465293298858045
Training model #14 ROEM: 0.1003615999313152
Training model #15 ROEM: 0.1008203990074308
Training model #16 ROEM: 0.0875392856326891
Training model #17 ROEM: 0.09363988542951324
Training model #18 ROEM: 0.09732934410211465
Training model #19 ROEM: 0.09712594484135964
Training model #20 ROEM: 0.08690172940195626
Training model #21 ROEM: 0.0934995323916171
Training model #22 ROEM: 0.0

In [31]:
# Find the index of the smallest ROEM
i = np.argmin(ROEMS)
print("Index of smallest ROEM:", i)

# Find ith element of ROEMS
print("Smallest ROEM: ", ROEMS[i])

Index of smallest ROEM: 152
Smallest ROEM:  0.04733565274244826


## Exercises

### Running a cross-validated implicit ALS model

Now that we have several ALS models, each with a different set of hyperparameter values, we can train them on a training portion of the msd dataset using cross validation, and then run them on a test set of data and evaluate how well each one performs using the ROEM function discussed earlier. Unfortunately, this takes too much time for this exercise, so it has been done separately. But for your reference you can evaluate your model_list using the following loop (we are using the msd dataset in this case).

For purposes of walking you through the steps, the test predictions for 192 models have already been generated, and their ROEM has been calculated. They are found in the ROEMS list provided. Because a list isn't unique to Pyspark, and because numpy works really well with lists, we're going to use numpy here. Follow the instructions below to find the best ROEM and the model that provided it.

In [32]:
# Split the data into training and test sets
(training, test) = msd.randomSplit([0.8, 0.2])

#Building 5 folds within the training set.
train1, train2, train3, train4, train5 = training.randomSplit([0.2, 0.2, 0.2, 0.2, 0.2], seed = 1)
fold1 = train2.union(train3).union(train4).union(train5)
fold2 = train3.union(train4).union(train5).union(train1)
fold3 = train4.union(train5).union(train1).union(train2)
fold4 = train5.union(train1).union(train2).union(train3)
fold5 = train1.union(train2).union(train3).union(train4)

foldlist = [(fold1, train1), (fold2, train2), (fold3, train3), (fold4, train4), (fold5, train5)]

ROEMS = []
# Loops through all models and all folds
for model in model_list:
    for ft_pair in foldlist:
        # Fits model to fold within training data
        fitted_model = model.fit(ft_pair[0])
        # Generates predictions using fitted_model on respective CV test data
        predictions = fitted_model.transform(ft_pair[1])
        # Generates and prints a ROEM metric CV test data
        r = roem(spark, predictions, 'userId', 'num_plays')
        print ("ROEM: ", r)
    # Fits model to all of training data and generates preds for test data
    v_fitted_model = model.fit(training)
    v_predictions = v_fitted_model.transform(test)
    v_ROEM = roem(spark, v_predictions, 'userId', 'num_plays')
    # Adds validation ROEM to ROEM list
    ROEMS.append(v_ROEM)
    print ("Validation ROEM: ", v_ROEM)

ROEM:  0.08617143998700197


ROEM:  0.08837538667608544


ROEM:  0.08583484653140143


ROEM:  0.0877753310739863


ROEM:  0.0886387865044038
Validation ROEM:  0.08678498620088304


ROEM:  0.08922609752153714


ROEM:  0.09729879663887452


ROEM:  0.09382281883971622
ROEM:  0.09383237432420558


ROEM:  0.09175231404624233
Validation ROEM:  0.09206651503495351


ROEM:  0.09340372178520179
ROEM:  0.09923608132093809


ROEM:  0.09996169302311418


ROEM:  0.10131481402099253


ROEM:  0.09918578895982279
Validation ROEM:  0.09764225072719901


ROEM:  0.09884698729267681
ROEM:  0.10145065349833773


ROEM:  0.1062097432403951
ROEM:  0.10599039885643509


ROEM:  0.10489198480003074
Validation ROEM:  0.10118575855518805
ROEM:  0.08553496453961307
ROEM:  0.08762839497160474
ROEM:  0.08495741461512235
ROEM:  0.08771657312263306
ROEM:  0.08777772268318115
Validation ROEM:  0.08604091005397228
ROEM:  0.088248556428937
ROEM:  0.09667168042486556


ROEM:  0.09296279308642807


ROEM:  0.09166698005556706


ROEM:  0.09115049123747115
Validation ROEM:  0.09091206423481953


ROEM:  0.09135825622142833


ROEM:  0.09819130626549376
ROEM:  0.09903012180565034
ROEM:  0.09992313320368497


ROEM:  0.09774482305327797
Validation ROEM:  0.09680596507205101
ROEM:  0.09711321897946054
ROEM:  0.09934751979403293


ROEM:  0.10532846142371495


ROEM:  0.10427888625350536


ROEM:  0.10367485863986069
Validation ROEM:  0.10041399590133007


ROEM:  0.08469427675471097
ROEM:  0.08721064783554591
ROEM:  0.08422112751294447
ROEM:  0.08760411323183047


ROEM:  0.0869421521216779
Validation ROEM:  0.08578965692957162


ROEM:  0.08712574888376826
ROEM:  0.09584503522892397


ROEM:  0.09223505945651289


ROEM:  0.08974537828378548
ROEM:  0.0906086405518309
Validation ROEM:  0.08939744142283139


ROEM:  0.08947293500923498


ROEM:  0.09656572119614325


ROEM:  0.09833276331496797
ROEM:  0.09852073405126564


ROEM:  0.09674840147787288
Validation ROEM:  0.09515625786762673


ROEM:  0.09650569080019432


ROEM:  0.09835916387250507


ROEM:  0.10465408896138337


ROEM:  0.10397635609001579


ROEM:  0.10284528264217568
Validation ROEM:  0.09865829790148559


ROEM:  0.08488731957847104


ROEM:  0.08995864444487224


ROEM:  0.08316086409784666
ROEM:  0.08949291184780482
ROEM:  0.08970712573604765
Validation ROEM:  0.08941151348919443


ROEM:  0.08816709494941491
ROEM:  0.09338807046623118
ROEM:  0.09125982834635267


ROEM:  0.0915294283559108
ROEM:  0.0886014298079181
Validation ROEM:  0.09243138805993611


ROEM:  0.09019468849279055


ROEM:  0.10004392353334317
ROEM:  0.09457072717971567


ROEM:  0.09798645651248701
ROEM:  0.09353859734875382
Validation ROEM:  0.09631899433027395


ROEM:  0.09571493281991034


ROEM:  0.0996005906592521


ROEM:  0.09962256190570148
ROEM:  0.10125807218710683


ROEM:  0.09841442093439752
Validation ROEM:  0.09888684064897747


ROEM:  0.08393524883233629


ROEM:  0.08706381284554358


ROEM:  0.08214678157842255
ROEM:  0.08839523207551978
ROEM:  0.08949886774665976
Validation ROEM:  0.0874467312251149


ROEM:  0.08743899941300581


ROEM:  0.09262783624876697


ROEM:  0.08909121909333717


ROEM:  0.0900821360716411
ROEM:  0.08832611942178062
Validation ROEM:  0.09012291769534005


ROEM:  0.0893121546717965


ROEM:  0.0983382716303877


ROEM:  0.09198112205896718
ROEM:  0.09604191579883424


ROEM:  0.09233375980534098
Validation ROEM:  0.09368799120182832


ROEM:  0.09394116467308852
ROEM:  0.0971929802794813


ROEM:  0.0983627900553688
ROEM:  0.09979790537211579


ROEM:  0.09647738682084209
Validation ROEM:  0.09700262026464845


ROEM:  0.08350519208058557
ROEM:  0.08651142409314314


ROEM:  0.08154561436810882
ROEM:  0.08721989960694757


ROEM:  0.08822894972566708
Validation ROEM:  0.08615536594660844


ROEM:  0.0859378172830633
ROEM:  0.091622599221846
ROEM:  0.08859671681722853
ROEM:  0.08878353096724294


ROEM:  0.08898085236600539
Validation ROEM:  0.09003673893231424
ROEM:  0.08698178743446716
ROEM:  0.09600277807391117


ROEM:  0.0903917037939171


ROEM:  0.09498314810662613


ROEM:  0.09086460986793739
Validation ROEM:  0.09184803076375726


ROEM:  0.09302383177599377
ROEM:  0.09513097282767906


ROEM:  0.09699742403614695
ROEM:  0.0987889544236047


ROEM:  0.09613136571386321
Validation ROEM:  0.09667819339758385


ROEM:  0.08374404397254154


ROEM:  0.08797164414306521


ROEM:  0.08331530752013391


ROEM:  0.08951416487349742


ROEM:  0.0894574534777273
Validation ROEM:  0.09073794008414576


ROEM:  0.08745265142318044


ROEM:  0.09309441926606773


ROEM:  0.08947829178623495
ROEM:  0.09119587643655744


ROEM:  0.08974791999609158
Validation ROEM:  0.09513864415161302
ROEM:  0.08967126883529812


ROEM:  0.09750117621048768


ROEM:  0.09075597795875881


ROEM:  0.09783098601152398
ROEM:  0.09213820406258277
Validation ROEM:  0.09481720223350887
ROEM:  0.09477687847729871


ROEM:  0.09799925856777165
ROEM:  0.0954967587998849


ROEM:  0.0999291801220139


ROEM:  0.09578912389469467
Validation ROEM:  0.0979029495294047


ROEM:  0.08253439128725178


ROEM:  0.08637048483283868
ROEM:  0.08204966941470605


ROEM:  0.08715207303601905


ROEM:  0.08915617996440386
Validation ROEM:  0.08918934487107727


ROEM:  0.0865112144829407


ROEM:  0.09137972498811905


ROEM:  0.08797062602867957


ROEM:  0.08897089923103828


ROEM:  0.08927973162635885
Validation ROEM:  0.0907666213610309


ROEM:  0.08934749937149297


ROEM:  0.09565550265995361
ROEM:  0.0902031067715068


ROEM:  0.09639842749121502


ROEM:  0.09110654881347752
Validation ROEM:  0.09332830684781944


ROEM:  0.09158334963350022
ROEM:  0.09762955624965883


ROEM:  0.09417339717238268


ROEM:  0.09912065700379497


ROEM:  0.09561346902759382
Validation ROEM:  0.09853481940257626


ROEM:  0.08287363281845257
ROEM:  0.08553701043696554


ROEM:  0.08109346711721752


ROEM:  0.08625923547857338


ROEM:  0.08796291050418967
Validation ROEM:  0.0871131868945286


ROEM:  0.08567400948392795
ROEM:  0.09055563880797549


ROEM:  0.08789705550909416


ROEM:  0.08851316962896688
ROEM:  0.08887853754267842
Validation ROEM:  0.08944532711952946


ROEM:  0.0867732263807401


ROEM:  0.09367160733281445


ROEM:  0.08940957062234664


ROEM:  0.09674732679481056
ROEM:  0.08853301193763513
Validation ROEM:  0.09145264606530477
ROEM:  0.09086441531805133


ROEM:  0.09529607193567625


ROEM:  0.09305652793668943
ROEM:  0.09847985433150123
ROEM:  0.09529815876410448
Validation ROEM:  0.09708248866002056


ROEM:  0.08312153494219336
ROEM:  0.08830339141856625


ROEM:  0.08289234447950296


ROEM:  0.0883142405394698


ROEM:  0.08892444631394626
Validation ROEM:  0.09156062231526665


ROEM:  0.0869117778400156


ROEM:  0.09254692094804039


ROEM:  0.09077520881163108
ROEM:  0.09170676701076765


ROEM:  0.09047588460860849
Validation ROEM:  0.09426817847109663


ROEM:  0.09018221971301317


ROEM:  0.09604847772017479


ROEM:  0.09158753472190126


ROEM:  0.09796755902321212
ROEM:  0.0910415315469979
Validation ROEM:  0.0945612864230796


ROEM:  0.09324997541150273


ROEM:  0.09666941846361486


ROEM:  0.09283818467012304


ROEM:  0.09927698028578996


ROEM:  0.09481632573982654
Validation ROEM:  0.09866003090165901


ROEM:  0.08233520218930654


ROEM:  0.08749580760027922


ROEM:  0.08219087037475205


ROEM:  0.08623158496662527


ROEM:  0.08789057198912092
Validation ROEM:  0.09130582754200306


ROEM:  0.0869674730868039


ROEM:  0.08991858033435621


ROEM:  0.08788752325845452


ROEM:  0.09057116316727191


ROEM:  0.08976460609105852
Validation ROEM:  0.0912630385038341


ROEM:  0.08979099601263357


ROEM:  0.09366452074985317


ROEM:  0.09004675468099507


ROEM:  0.09764703062167154


ROEM:  0.09123574661578884
Validation ROEM:  0.09409569236514577


ROEM:  0.09170596140119361


ROEM:  0.09662673382845617


ROEM:  0.09254736257925966


ROEM:  0.09879077537955537


ROEM:  0.09549797072373536
Validation ROEM:  0.09646561333754637


ROEM:  0.08198669256212932


ROEM:  0.08585481662581881


ROEM:  0.08071545336470959


ROEM:  0.08615903234899012


ROEM:  0.08711962552921847
Validation ROEM:  0.08744146145769473


ROEM:  0.08544801323510479


ROEM:  0.08922416800717423


ROEM:  0.0881079780084663


ROEM:  0.08970975624481865


ROEM:  0.08923706834028242
Validation ROEM:  0.09014324646323936
ROEM:  0.08627094187267394


ROEM:  0.09306036290743637


ROEM:  0.08961512631291088


ROEM:  0.09668214249782732


ROEM:  0.0884650073147943
Validation ROEM:  0.0923667330121221


ROEM:  0.09026165149253228
ROEM:  0.0979824420486786


ROEM:  0.09185507895366903


ROEM:  0.09829971729105241


ROEM:  0.0953659238552808
Validation ROEM:  0.09397088825504939


ROEM:  0.06849441945097012


ROEM:  0.07149692707428178
ROEM:  0.07086726435539133


ROEM:  0.07316760216589259


ROEM:  0.07137150181500894
Validation ROEM:  0.07001573289658256
ROEM:  0.07269629509706094


ROEM:  0.07573160211189112


ROEM:  0.07551400160985573


ROEM:  0.07671457217205149


ROEM:  0.07475466109241398
Validation ROEM:  0.07214403908264008


ROEM:  0.07914961193910373
ROEM:  0.07897707806844675


ROEM:  0.07677608703527151


ROEM:  0.07945754389916598


ROEM:  0.07764232819074408
Validation ROEM:  0.07474882611712741


ROEM:  0.08054946873744667


ROEM:  0.08188280890437503


ROEM:  0.07964104008293874


ROEM:  0.08157809548620688


ROEM:  0.08042770315382206
Validation ROEM:  0.07589064208115703


ROEM:  0.06730215967509741
ROEM:  0.06971597228359395


ROEM:  0.0673178041283671


ROEM:  0.07057099559403483


ROEM:  0.0698694726592084
Validation ROEM:  0.06905256514616294


ROEM:  0.07203903716986637


ROEM:  0.07345796635031293


ROEM:  0.07368974313575778


ROEM:  0.07447899477454588


ROEM:  0.07255390053528774
Validation ROEM:  0.07105910394466289
ROEM:  0.07764434944212437


ROEM:  0.07755436391726446


ROEM:  0.07475008109575525


ROEM:  0.07709842140450861


ROEM:  0.07604429591008176
Validation ROEM:  0.07359972500714929


ROEM:  0.07912260070685612


ROEM:  0.07906432467294848
ROEM:  0.07817095499006611


ROEM:  0.07978352259222572


ROEM:  0.07808334005842633
Validation ROEM:  0.07523113475346996


ROEM:  0.06654422707619106


ROEM:  0.06821105096022553


ROEM:  0.06606441965637062


ROEM:  0.06986323226373106


ROEM:  0.06935894251128248
Validation ROEM:  0.06754877766901915


ROEM:  0.07092332421478521


ROEM:  0.07225330933914365


ROEM:  0.07210001119640856


ROEM:  0.07272653414124647


ROEM:  0.0713305358742568
Validation ROEM:  0.06951278075934093
ROEM:  0.0758667347681628


ROEM:  0.07605610710873087


ROEM:  0.07390690544455299


ROEM:  0.07521197503875611


ROEM:  0.07520055908345533
Validation ROEM:  0.07214432005109472


ROEM:  0.07830307236568138


ROEM:  0.07716235722152524


ROEM:  0.07707590911725326


ROEM:  0.07824923060106505


ROEM:  0.07692003875824123
Validation ROEM:  0.07519498400737513


ROEM:  0.066692361430224


ROEM:  0.071739887643362


ROEM:  0.06931087501165728


ROEM:  0.07244472301138148


ROEM:  0.07140073654259844
Validation ROEM:  0.0698320100332564


ROEM:  0.07088496420069924


ROEM:  0.07370221715293415


ROEM:  0.07165268994804343


ROEM:  0.07533515311473443


ROEM:  0.0736192289882092
Validation ROEM:  0.07007906500071949


ROEM:  0.07600508478607622
ROEM:  0.0761023680403978


ROEM:  0.07318401257889717


ROEM:  0.07685694267291196


ROEM:  0.07588656426141972
Validation ROEM:  0.0735118133428047


ROEM:  0.0757044787506528


ROEM:  0.0784878585096378


ROEM:  0.07778498949203441


ROEM:  0.07868841451564258


ROEM:  0.07804455085415697
Validation ROEM:  0.07524590348806541


ROEM:  0.0650019312793189


ROEM:  0.0704574503954369
ROEM:  0.06493933920542067


ROEM:  0.07076589881509936


ROEM:  0.07014909141553163
Validation ROEM:  0.06851494777620483


ROEM:  0.0714526060968031


ROEM:  0.0720058583635069
ROEM:  0.06977613176337326
ROEM:  0.07495212829208502
ROEM:  0.07054402146785423
Validation ROEM:  0.06962915103230512


ROEM:  0.07508767401024445


ROEM:  0.07359373875523764


ROEM:  0.07131765855862254


ROEM:  0.07467311662128502


ROEM:  0.07391341664085169
Validation ROEM:  0.07099120207153889


ROEM:  0.07389571546071502


ROEM:  0.07542946975863818


ROEM:  0.07555772451783284


ROEM:  0.07737235845344007


ROEM:  0.07559623369120091
Validation ROEM:  0.0729349852987667


ROEM:  0.06500014504071151


ROEM:  0.06950975779252432


ROEM:  0.06433861546068609


ROEM:  0.06939046904487811


ROEM:  0.06890885782296705
Validation ROEM:  0.06727929528343242


ROEM:  0.06986669737537334


ROEM:  0.07173872398717805


ROEM:  0.06767694491097613


ROEM:  0.0723399646920128


ROEM:  0.07018756668867042
Validation ROEM:  0.06787465503188501


ROEM:  0.07354223245964714


ROEM:  0.07132482177568712


ROEM:  0.07044862379249224


ROEM:  0.07336500520900428
ROEM:  0.07210597812030033
Validation ROEM:  0.07045383163919516


ROEM:  0.07293447101435213


ROEM:  0.07336550913515412
ROEM:  0.07376483425094614


ROEM:  0.07536352159262288


ROEM:  0.07400130747264454
Validation ROEM:  0.07142740426835792


ROEM:  0.06674273104794452


ROEM:  0.07110379318696519


ROEM:  0.06826785269053642


ROEM:  0.07304644711781035


ROEM:  0.07164037556672043
Validation ROEM:  0.07057385957212152


ROEM:  0.07026754465769718


ROEM:  0.0733504472709073


ROEM:  0.06914374585873861


ROEM:  0.07484336842640235


ROEM:  0.07308120888861827
Validation ROEM:  0.07016034327986809


ROEM:  0.07448449767999657


ROEM:  0.07450643558088736


ROEM:  0.07142375601852238


ROEM:  0.07542972165711108


ROEM:  0.07525540653978299
Validation ROEM:  0.07348612814313499


ROEM:  0.07386464872649252


ROEM:  0.07645948213950639


ROEM:  0.0777732804663009


ROEM:  0.07726812557541592


ROEM:  0.07645827268733393
Validation ROEM:  0.07510425774553467


ROEM:  0.06460998196747339


ROEM:  0.07014242160072034


ROEM:  0.06441838846442377


ROEM:  0.0697172994740286


ROEM:  0.07002736726208117
Validation ROEM:  0.06859692669203779


ROEM:  0.07087739205267661
ROEM:  0.07151196112767431


ROEM:  0.06776564576223419


ROEM:  0.07441042086187477


ROEM:  0.0704183440096632
Validation ROEM:  0.0675870684563409


ROEM:  0.07364094565810836


ROEM:  0.07358671541582115
ROEM:  0.0707490772521151
ROEM:  0.07363100884444138
ROEM:  0.07306501093243091
Validation ROEM:  0.06976396851969748


ROEM:  0.07084398963897781


ROEM:  0.07239208283004872


ROEM:  0.0738855604417757


ROEM:  0.07649645955024247


ROEM:  0.0740229009786734
Validation ROEM:  0.07277854903965396


ROEM:  0.06513276910598834


ROEM:  0.06937343120316146


ROEM:  0.06351328596465355


ROEM:  0.06914163419329541


ROEM:  0.06904229821690744
Validation ROEM:  0.06744877507313388


ROEM:  0.0694154321860627


ROEM:  0.07017109734986005


ROEM:  0.06648648258218019


ROEM:  0.07326918303575874


ROEM:  0.06947271031359731
Validation ROEM:  0.06650327781434558


ROEM:  0.07235750920570268


ROEM:  0.06986353008112799


ROEM:  0.07034996399055576


ROEM:  0.07293626575288302


ROEM:  0.07070904758731802
Validation ROEM:  0.06954226930668121


ROEM:  0.07086121260271346


ROEM:  0.07218081240906898


ROEM:  0.07289864043052187


ROEM:  0.07534880806072206


ROEM:  0.07236374891979974
Validation ROEM:  0.07152704179534217


ROEM:  0.06807183335062475


ROEM:  0.07318336395861996


ROEM:  0.0679908892011976


ROEM:  0.07280953487074722


ROEM:  0.0716246006382442
Validation ROEM:  0.07048263992899727


ROEM:  0.07089113369613781


ROEM:  0.07272906897996828


ROEM:  0.06821894329711954


ROEM:  0.07596769447728306


ROEM:  0.07293195625328874
Validation ROEM:  0.07011974355762916


ROEM:  0.0742312658579994


ROEM:  0.0736378998216343
ROEM:  0.0718882377447684


ROEM:  0.07583650733125799


ROEM:  0.07463708229641429
Validation ROEM:  0.07206876823086308


ROEM:  0.0726125085789092


ROEM:  0.0751060174122215


ROEM:  0.07605683935783987


ROEM:  0.07695204270382408


ROEM:  0.07586209711271225
Validation ROEM:  0.07407753370509597


ROEM:  0.06575902853163937


ROEM:  0.0700896876451916


ROEM:  0.06394686806764498


ROEM:  0.068935439490584


ROEM:  0.07101824535457622
Validation ROEM:  0.06846592249427659


ROEM:  0.0710518240589525


ROEM:  0.07047631497443664


ROEM:  0.06817559161923695


ROEM:  0.07429410500659137


ROEM:  0.07107689175200105
Validation ROEM:  0.06716535115618305


ROEM:  0.07298104600846354


ROEM:  0.07244460389322122


ROEM:  0.07098579792640254


ROEM:  0.07309911206564593


ROEM:  0.07270185963534971
Validation ROEM:  0.06883340403811773


ROEM:  0.07050199301781994


ROEM:  0.07195236467514975


ROEM:  0.07285364380850783


ROEM:  0.07637405220947915


ROEM:  0.07245157920500285
Validation ROEM:  0.07287271794474437


ROEM:  0.06522297335551112


ROEM:  0.06881169070429186
ROEM:  0.06348324317765155


ROEM:  0.06900506073523612


ROEM:  0.06959734749571103
Validation ROEM:  0.06806162833270475


ROEM:  0.06916423524347758
ROEM:  0.06939234327824205


ROEM:  0.06678936165854472


ROEM:  0.07342557555700203


ROEM:  0.06936470180130126
Validation ROEM:  0.06619751831400643


ROEM:  0.07242038123110123


ROEM:  0.07050015867167139


ROEM:  0.07102867222921967


ROEM:  0.07260001284847768


ROEM:  0.07102451321626244
Validation ROEM:  0.06829920179649421


ROEM:  0.07038897281739208


ROEM:  0.071847455490439
ROEM:  0.0718059231532587


ROEM:  0.07496120676288544
ROEM:  0.07182958860688897
Validation ROEM:  0.07135452786692334
ROEM:  0.0599330964572337
ROEM:  0.06146392865619323


ROEM:  0.06005037847234737


ROEM:  0.06224255110468043


ROEM:  0.06514440698396447
Validation ROEM:  0.058507898311697136


ROEM:  0.06047111293533437


ROEM:  0.06212585990846027


ROEM:  0.0651322715376972


ROEM:  0.07009901376507553


ROEM:  0.06680230149669668
Validation ROEM:  0.06211792004916256


ROEM:  0.06275571833084266


ROEM:  0.06493951927571845


ROEM:  0.06651983010582094


ROEM:  0.07273844782874567


ROEM:  0.0679102939636465
Validation ROEM:  0.06348815364426216
ROEM:  0.06532189026951937


ROEM:  0.07060853262268403


ROEM:  0.070062633013418


ROEM:  0.07448287450150709


ROEM:  0.07049957673735059
Validation ROEM:  0.06409664203303121


ROEM:  0.058482963584374964


ROEM:  0.05910887053437356


ROEM:  0.059656478637053215


ROEM:  0.06128934808532393


ROEM:  0.062042001811429055
Validation ROEM:  0.05766655702071457


ROEM:  0.05798594597682218


ROEM:  0.06047159732989644


ROEM:  0.06257045579337664


ROEM:  0.06605528496975663


ROEM:  0.06535048384420801
Validation ROEM:  0.05888643870717509


ROEM:  0.06093339286342318


ROEM:  0.06391253659438312


ROEM:  0.06405175676627464


ROEM:  0.07126165594832437


ROEM:  0.06580252580591357
Validation ROEM:  0.06140907683221327


ROEM:  0.06264584686892982


ROEM:  0.06635278682796689


ROEM:  0.06715584511661955


ROEM:  0.07183987445266628


ROEM:  0.06702713857590104
Validation ROEM:  0.06188259748662725


ROEM:  0.057633389037291864


ROEM:  0.05781339764680457


ROEM:  0.05960876686173768
ROEM:  0.06124307798309739
ROEM:  0.060503301651438704
Validation ROEM:  0.0564770817716647


ROEM:  0.05730272424932455


ROEM:  0.06054597780321689


ROEM:  0.06108020051220962


ROEM:  0.06441623888886204


ROEM:  0.06377374236132849
Validation ROEM:  0.05710969160130639
ROEM:  0.060283600856045685


ROEM:  0.06101244361435904


ROEM:  0.06293844015323384


ROEM:  0.06787892498104964


ROEM:  0.06512050097902253
Validation ROEM:  0.06026587599898471


ROEM:  0.06239529231995784


ROEM:  0.0631627639893063
ROEM:  0.0641835260964594


ROEM:  0.07040853414571534


ROEM:  0.06513050994189348
Validation ROEM:  0.06128016624816688


ROEM:  0.0584777238279347
ROEM:  0.060024324501095384


ROEM:  0.05983634024589644


ROEM:  0.061809216118923586


ROEM:  0.06403112260342825
Validation ROEM:  0.059786985066185024


ROEM:  0.05995803096317418


ROEM:  0.061339354712000686


ROEM:  0.06477244108529044
ROEM:  0.06754195592691886


ROEM:  0.06616517436172836
Validation ROEM:  0.058552277145861066


ROEM:  0.06066712582874138


ROEM:  0.06202686523450442


ROEM:  0.06555310104339628
ROEM:  0.07006312112060044


ROEM:  0.06689625331301254
Validation ROEM:  0.06152199806562477


ROEM:  0.06398858835229944
ROEM:  0.06615370285715788
ROEM:  0.06587289615663382


ROEM:  0.07252503537166911
ROEM:  0.06861049994732649
Validation ROEM:  0.06339926437131896


ROEM:  0.05789560563048003


ROEM:  0.05701242329083001


ROEM:  0.058924170705467314


ROEM:  0.06221128565871038


ROEM:  0.06250429711552463
Validation ROEM:  0.05856299173667054


ROEM:  0.058994669497949265


ROEM:  0.05891048770276689


ROEM:  0.062168413058362425


ROEM:  0.06512686294935038


ROEM:  0.06506216305447254
Validation ROEM:  0.05825718330149835


ROEM:  0.05911442027298822


ROEM:  0.06132204315302137


ROEM:  0.06322491912874098


ROEM:  0.06840930415098004


ROEM:  0.066327926572162
Validation ROEM:  0.059793124039794865


ROEM:  0.06157824677145308


ROEM:  0.06189827548162923


ROEM:  0.06201868777523589


ROEM:  0.06982813652210017


ROEM:  0.06642707917498111
Validation ROEM:  0.06114294145678215
ROEM:  0.05843169137946043


ROEM:  0.05542209677090418


ROEM:  0.05846280614404518


ROEM:  0.061285121585212626


ROEM:  0.06089134687631919
Validation ROEM:  0.05713587735468085


ROEM:  0.0594030072884634


ROEM:  0.05974772299749142


ROEM:  0.06072154429461772


ROEM:  0.06373662805611166


ROEM:  0.06439376037109612
Validation ROEM:  0.05715942272111182


ROEM:  0.058414834402111766


ROEM:  0.06125590847032469


ROEM:  0.061371900629031646


ROEM:  0.06705907731243975


ROEM:  0.06483371106154309
Validation ROEM:  0.058889114958749086


ROEM:  0.06147982611913445


ROEM:  0.0603544425756023


ROEM:  0.06231073451834775


ROEM:  0.06775591281037886
ROEM:  0.0659158700572589
Validation ROEM:  0.05997854779492761


ROEM:  0.05892408672079893


ROEM:  0.05949021915851035


ROEM:  0.0588040313202961


ROEM:  0.06374030326190326


ROEM:  0.06470137945598538
Validation ROEM:  0.06035131910436517


ROEM:  0.06082277166699591


ROEM:  0.06168403339119556


ROEM:  0.06358601971048362


ROEM:  0.06748735196545885


ROEM:  0.06638348359494085
Validation ROEM:  0.05985692945794254


ROEM:  0.06022899839494673


ROEM:  0.0616450259455188
ROEM:  0.06558593019254988


ROEM:  0.07029945097534473


ROEM:  0.0668406537300308
Validation ROEM:  0.06022131488433199


ROEM:  0.061547817112922634


ROEM:  0.06605318486411581


ROEM:  0.06447880993890912


ROEM:  0.07140934900346053


ROEM:  0.06810380833297719
Validation ROEM:  0.06348223133200873
ROEM:  0.058353255233571696


ROEM:  0.057444967425511785


ROEM:  0.05864333828774179


ROEM:  0.0631464867507409
ROEM:  0.06290909089685796
Validation ROEM:  0.05820043581291026


ROEM:  0.059330466011993904


ROEM:  0.058864828293679866


ROEM:  0.06281089181622992


ROEM:  0.0649363759089195


ROEM:  0.06402969214044076
Validation ROEM:  0.0586026763414957


ROEM:  0.05895418603591483


ROEM:  0.06061979711822713


ROEM:  0.06287518471845784


ROEM:  0.06708328474151572


ROEM:  0.06599004876320536
Validation ROEM:  0.059592907914348466


ROEM:  0.06017227640201709


ROEM:  0.06207373685316676


ROEM:  0.061518709979773094


ROEM:  0.06860032008790933


ROEM:  0.06533636368550591
Validation ROEM:  0.06093857631444795


ROEM:  0.059740798216070076


ROEM:  0.055878790652392304


ROEM:  0.05859827247443271


ROEM:  0.062293956330389595


ROEM:  0.061922373802679384
Validation ROEM:  0.056896683893230146


ROEM:  0.05937704840381588


ROEM:  0.059745375832270596


ROEM:  0.06113884282377892


ROEM:  0.06408917938320405


ROEM:  0.06527162486202136
Validation ROEM:  0.057685441180443116
ROEM:  0.058658917083986656
ROEM:  0.06165999638734249


ROEM:  0.061309218512251476


ROEM:  0.06667978014786917


ROEM:  0.06465370984571256
Validation ROEM:  0.058347064588287194


ROEM:  0.06045693045498914


ROEM:  0.059706219509571355


ROEM:  0.06190739601210307


ROEM:  0.06763827125262183


ROEM:  0.06567095610339581
Validation ROEM:  0.05935975056262684


ROEM:  0.059197965360642764


ROEM:  0.059091698713039764


ROEM:  0.05914462736449925


ROEM:  0.06460044097455256
ROEM:  0.0642957324159114
Validation ROEM:  0.060249984062521564


ROEM:  0.06026910568575324


ROEM:  0.06087033503609402


ROEM:  0.0634726791337921


ROEM:  0.06620525053261496


ROEM:  0.06679540256320032
Validation ROEM:  0.060416884118524104


ROEM:  0.06127330921208848


ROEM:  0.061133690842179546


ROEM:  0.0651707797164681


ROEM:  0.069470708117442
ROEM:  0.06581412404594064
Validation ROEM:  0.05931912311754482


ROEM:  0.061425006438554225


ROEM:  0.06582563393012122


ROEM:  0.06411036899769025


ROEM:  0.07071437495276904


ROEM:  0.06751505901001371
Validation ROEM:  0.06310512887483043


ROEM:  0.05838175510859693


ROEM:  0.0569116012388312


ROEM:  0.059053768163580794


ROEM:  0.06339795887400335


ROEM:  0.06298841856174643
Validation ROEM:  0.05847044298806712


ROEM:  0.059242022588449655


ROEM:  0.059288057688091686


ROEM:  0.06206623145485083


ROEM:  0.06406209185269555


ROEM:  0.06470395626992967
Validation ROEM:  0.058565922929348965


ROEM:  0.0602179324128204


ROEM:  0.06081786018906827


ROEM:  0.06329572195024415


ROEM:  0.06616494600286173


ROEM:  0.06564293636100188
Validation ROEM:  0.05898618857139784


ROEM:  0.06003623092004336


ROEM:  0.06132689205999097


ROEM:  0.06083591497054892


ROEM:  0.06820453279597385


ROEM:  0.06517389943349197
Validation ROEM:  0.0609306377450533


ROEM:  0.06097140305997995


ROEM:  0.056388630967477815


ROEM:  0.058837657424454196


ROEM:  0.0626193225855281


ROEM:  0.06204705879703992
Validation ROEM:  0.057051857996645736


ROEM:  0.05926554941234011


ROEM:  0.06028696706679801


ROEM:  0.06121519640363653


ROEM:  0.06412827641328682


ROEM:  0.06544751858766731
Validation ROEM:  0.05814483547997177


ROEM:  0.05877278463872816


ROEM:  0.062136525987950254


ROEM:  0.0597974198091546


ROEM:  0.06636939796850892


ROEM:  0.06486946760661183
Validation ROEM:  0.05799195495001348


ROEM:  0.0601062779841199


ROEM:  0.06007644163683178


ROEM:  0.061390271305681254


ROEM:  0.06806504780861779


ROEM:  0.06539410001829829
Validation ROEM:  0.05945166435997864


ROEM:  0.05239796699087031


ROEM:  0.05705259178513576


ROEM:  0.054225733421052025


ROEM:  0.05922106257132221


ROEM:  0.05704997886360201
Validation ROEM:  0.05010519476024194


ROEM:  0.05491305643307378


ROEM:  0.058973021112995634


ROEM:  0.0534558626365575


ROEM:  0.060905284288631775


ROEM:  0.058220558609099374
Validation ROEM:  0.05366738089240896


ROEM:  0.05749406323666367


ROEM:  0.05954427941777165


ROEM:  0.057018877782163274


ROEM:  0.06083535797772595


ROEM:  0.06003633437335608
Validation ROEM:  0.054709797374302246


ROEM:  0.060119210990065924


ROEM:  0.062053038453363595


ROEM:  0.05948203631532353


ROEM:  0.06322448423763564


ROEM:  0.06209608068797205
Validation ROEM:  0.05748706921210576


ROEM:  0.05158773224531147


ROEM:  0.05496313772839218


ROEM:  0.05377328547719537


ROEM:  0.05724334574834988


ROEM:  0.055413905792648134
Validation ROEM:  0.04940505023215477


ROEM:  0.0538966195007317


ROEM:  0.058338288059712806


ROEM:  0.05207638480473757


ROEM:  0.05981749269281271


ROEM:  0.05604096427939102
Validation ROEM:  0.051615679987268105


ROEM:  0.05529817656416648


ROEM:  0.057645109477889596


ROEM:  0.052825449498109855


ROEM:  0.05988712882327293


ROEM:  0.05810714063693332
Validation ROEM:  0.05263149759148976


ROEM:  0.05867942010114534


ROEM:  0.05965000193839492


ROEM:  0.057475374217493755


ROEM:  0.060598484824913146


ROEM:  0.05939673158334147
Validation ROEM:  0.05450509290942297


ROEM:  0.05171917573110131


ROEM:  0.05413366548762751


ROEM:  0.052672379214081286


ROEM:  0.05749865961850463


ROEM:  0.05480509052494088
Validation ROEM:  0.04914297738899683


ROEM:  0.052562366293488864


ROEM:  0.055505421137346314


ROEM:  0.05184999850067487


ROEM:  0.058958659894409425


ROEM:  0.055505560628770274
Validation ROEM:  0.05095301513736043


ROEM:  0.05461956293273687


ROEM:  0.05728155280463328


ROEM:  0.050902385214986065


ROEM:  0.05871715822146751


ROEM:  0.056801908113755274
Validation ROEM:  0.051545089245957375


ROEM:  0.056843174566176505


ROEM:  0.05804291542531458


ROEM:  0.05452277845362127


ROEM:  0.06029711523445132


ROEM:  0.05739378319264108
Validation ROEM:  0.05246729729464868


ROEM:  0.052780783914701423


ROEM:  0.05677993268819699


ROEM:  0.05454657236209129


ROEM:  0.058916591528614434


ROEM:  0.056854686334323105
Validation ROEM:  0.05046129459071487


ROEM:  0.05379845853436582
ROEM:  0.057335593860352585


ROEM:  0.051391345789655665


ROEM:  0.059926667232513386


ROEM:  0.05728512216345153
Validation ROEM:  0.053323569571403114


ROEM:  0.05577079317997983


ROEM:  0.05695398694809639


ROEM:  0.055202005596121456


ROEM:  0.05961035696605433


ROEM:  0.05933854050996136
Validation ROEM:  0.05377138896736909


ROEM:  0.05781076620477488


ROEM:  0.06117535339249486


ROEM:  0.05642076080309329


ROEM:  0.06142638227038403


ROEM:  0.059814890092371266
Validation ROEM:  0.0545005214068089


ROEM:  0.051786796426148966


ROEM:  0.05536572248683962


ROEM:  0.05326987068295435


ROEM:  0.05744021867488754


ROEM:  0.05640577809095526
Validation ROEM:  0.04990853846572004


ROEM:  0.05268335598350992


ROEM:  0.05689826336057494


ROEM:  0.05153180867342303


ROEM:  0.05933708699249885


ROEM:  0.0556969256374086
Validation ROEM:  0.05055867049742194


ROEM:  0.05459413040720273


ROEM:  0.05653816483657628


ROEM:  0.053307468937655436


ROEM:  0.058812872302046064


ROEM:  0.05713311003300052
Validation ROEM:  0.05266519578062024


ROEM:  0.057132090551018845


ROEM:  0.057900887721251036


ROEM:  0.05560489369375835


ROEM:  0.06052539824245526


ROEM:  0.0583787666731551
Validation ROEM:  0.05385734466864538


ROEM:  0.05165932983094836


ROEM:  0.05512940724916963


ROEM:  0.05233759839692143


ROEM:  0.05648089873361396


ROEM:  0.05537487222301937
Validation ROEM:  0.049110232501098576


ROEM:  0.051964510681153424


ROEM:  0.054178927525990815


ROEM:  0.052512669253479964
ROEM:  0.05772673902713216


ROEM:  0.0547195160231256
Validation ROEM:  0.05111663237095563


ROEM:  0.05371939974907224


ROEM:  0.055693756331284844


ROEM:  0.05080183030157639


ROEM:  0.05874627678082779


ROEM:  0.056247931126632414
Validation ROEM:  0.05152626619331321


ROEM:  0.05582914029342948


ROEM:  0.05657597004582063


ROEM:  0.05271132651094566


ROEM:  0.05908026295048959
ROEM:  0.05742287235460913
Validation ROEM:  0.05226721043333518


ROEM:  0.053535020478563525


ROEM:  0.05689664228196669


ROEM:  0.05448551199084777


ROEM:  0.059653694430192854


ROEM:  0.05742267762309282
Validation ROEM:  0.05097145288417213


ROEM:  0.05350510605633684


ROEM:  0.05701147248305899


ROEM:  0.05223097133214464


ROEM:  0.05970536175928732


ROEM:  0.05649666664648143
Validation ROEM:  0.0528167652311043


ROEM:  0.05630400770988021


ROEM:  0.056920775439272174


ROEM:  0.05487884299814265


ROEM:  0.05979131979445404


ROEM:  0.05848400261344193
Validation ROEM:  0.053404208132766126


ROEM:  0.05697581543050658


ROEM:  0.059963812989240924


ROEM:  0.05638773353648453


ROEM:  0.06107221736984123


ROEM:  0.05899818169543081
Validation ROEM:  0.053613684363381243


ROEM:  0.05183572726564856


ROEM:  0.05699797639375165
ROEM:  0.05318648731027015


ROEM:  0.05784900712239505


ROEM:  0.0564036264446535
Validation ROEM:  0.05074170346569431


ROEM:  0.053238106698146756


ROEM:  0.05583717519296769


ROEM:  0.05239407801220802


ROEM:  0.05830642709538012


ROEM:  0.05554904270500931
Validation ROEM:  0.05177870613862858


ROEM:  0.05478147459800452


ROEM:  0.056147653873895344


ROEM:  0.05444765219131293


ROEM:  0.05796013559482777


ROEM:  0.056614165209907606
Validation ROEM:  0.05262334232978582


ROEM:  0.05623021120297131


ROEM:  0.05645697552182134


ROEM:  0.05428959844333533


ROEM:  0.06097150684430919


ROEM:  0.057793120200636976
Validation ROEM:  0.05268036949168688


ROEM:  0.05280193012380983


ROEM:  0.05479885522745687


ROEM:  0.052221398928760175


ROEM:  0.05673533007217195


ROEM:  0.05539407299678601
Validation ROEM:  0.049251552190427525


ROEM:  0.05111144381824526


ROEM:  0.054315328773466606


ROEM:  0.052615548513836125


ROEM:  0.05755906190382258


ROEM:  0.05471898703616676
Validation ROEM:  0.05124713398784083


ROEM:  0.05430683970492732
ROEM:  0.05445880039904909


ROEM:  0.05148297286261259


ROEM:  0.05844195476431922


ROEM:  0.05696755527769938
Validation ROEM:  0.0509765300997309


ROEM:  0.0554221326908878


ROEM:  0.05652798867044578


ROEM:  0.05250671398348012


ROEM:  0.059228243356041745


ROEM:  0.057867256343615836
Validation ROEM:  0.05208319650999998


ROEM:  0.05429739433358803


ROEM:  0.05695147498344686


ROEM:  0.0554715179997275


ROEM:  0.06010909410284255


ROEM:  0.05790830631465033
Validation ROEM:  0.05170081695526577


ROEM:  0.05415734885026318


ROEM:  0.056407405218185845


ROEM:  0.05291096473020459


ROEM:  0.05993715191898928


ROEM:  0.05675015688961808
Validation ROEM:  0.05205632711441303


ROEM:  0.05620762277832538


ROEM:  0.05590788532728013


ROEM:  0.0545211844651951


ROEM:  0.0598332957402353


ROEM:  0.059445633521864286
Validation ROEM:  0.0533481155135179


ROEM:  0.057057440226869344


ROEM:  0.059278615798516815


ROEM:  0.05644822273046533


ROEM:  0.06061295723542736


ROEM:  0.059151116734398226
Validation ROEM:  0.05400361760864049


ROEM:  0.05263249195007428


ROEM:  0.05712358299846846


ROEM:  0.05357080976491893


ROEM:  0.05849228363939012


ROEM:  0.05634214768420832
Validation ROEM:  0.05117604462921908


ROEM:  0.05299104050792394


ROEM:  0.05627544433339322
ROEM:  0.05301057417585315


ROEM:  0.058468715926826065


ROEM:  0.05593944813320857
Validation ROEM:  0.05218409594830242
ROEM:  0.05503001527389023


ROEM:  0.05496656553801523


ROEM:  0.05434956322863732


ROEM:  0.05848952388605504


ROEM:  0.056344301769365855
Validation ROEM:  0.05211556212526388


ROEM:  0.05568354142380422
ROEM:  0.05552570196998368


ROEM:  0.05456949885483665
ROEM:  0.06133647162033824


ROEM:  0.05792954761735865
Validation ROEM:  0.052579264968038594


ROEM:  0.053030912904051895


ROEM:  0.05437357829042285


ROEM:  0.05255007501485606
ROEM:  0.05716885344407789


ROEM:  0.0556309814859046
Validation ROEM:  0.05025571242466438
ROEM:  0.05141289678687604


ROEM:  0.054266749819779046


ROEM:  0.05221823518942011


ROEM:  0.05731597613252354


ROEM:  0.05475556341488934
Validation ROEM:  0.051249551327145496


ROEM:  0.05394595470239455


ROEM:  0.05369003265599861


ROEM:  0.051206287447104266


ROEM:  0.05817952598978772


ROEM:  0.05703874998154857
Validation ROEM:  0.05127915871335019


ROEM:  0.05513256409507222


ROEM:  0.055628900450711845


ROEM:  0.05196429069014226


ROEM:  0.059512920839545104


ROEM:  0.05734069251731454
Validation ROEM:  0.05180534913323652


> Last executed at 2023-05-05 20:02:54 in 1h 24m 53s

In [33]:
# Find the index of the smallest ROEM
i = np.argmin(ROEMS)
print(f'Smallest ROEM #{i}: {ROEMS[i]}')

Smallest ROEM #164: 0.049110232501098576


### Extracting parameters

In [34]:
# Extract the best_model
best_model = model_list[i]

# Extract the Rank
print ("Rank: ", best_model.getRank())

# Extract the MaxIter value
print ("MaxIter: ", best_model.getMaxIter())

# Extract the RegParam value
print ("RegParam: ", best_model.getRegParam())

# Extract the Alpha value
print ("Alpha: ", best_model.getAlpha())

Rank:  40
MaxIter:  20
RegParam:  0.15
Alpha:  20.0


> Great work. Looks like a low rank, a higher maxIter, a low regParam, and a medium-high alpha is keeping the ROEM low. Because some of these values are on the high and low ends of the values we tried, it would be worth adding some additional values to test in our hyperparameter values, and doing this step again. But for right now, you should understand the process.